crates/telio-relay/src/derp/http.rs

```diff
@@ -89,21 +94,67 @@ pub async fn connect_http_and_start(
     };
     let hostport = format!("{}:{}", hostname, port);
 
-    let socket = socket_pool.new_external_tcp_v4(Some(TcpParams {
-        keepalive_enable: Some(true),
-        keepalive_cnt: Some(TCP_KEEPALIVE_COUNT),
-        keepalive_idle: Some(TCP_KEEPALIVE_IDLE),
-        keepalive_intvl: Some(TCP_KEEPALIVE_INTERVAL),
-        nodelay_enable: Some(true),
-        user_timeout: Some(TCP_USER_TIMEOUT),
-        buf_size: SocketBufSizes {
-            tx_buf_size: Some(SOCK_BUF_SZ),
-            rx_buf_size: Some(SOCK_BUF_SZ),
-        },
-    }))?;
+    // First try to use derp v2, with poll keepalives it the feature is enabled
+    if derp_config.server_keepalives.poll_keepalive {
+        telio_log_debug!("Trying to connect to derp V2");
+        let socket = socket_pool.new_external_tcp_v4(Some(build_tcp_parameters(false)))?;
+        let stream = timeout(derp_config.timeout, socket.connect(ip)).await??;
+        let addr = PairAddr {
+            local: stream.local_addr()?,
+            remote: stream.peer_addr()?,
+        };
 
-    let stream = timeout(derp_config.timeout, socket.connect(ip)).await??;
+        let connection = match u.scheme() {
+            "http" => {
+                Box::pin(connect_and_start(
+                    stream,
+                    addr,
+                    derp_config.secret_key.clone(),
+                    derp_config.server_keepalives.clone(),
+                    &hostport,
+                    DerpVersion::V2,
+                ))
+                .await
+            }
+            _ => {
+                let config = if derp_config.use_built_in_root_certificates {
+                    let root_store: RootCertStore = TLS_SERVER_ROOTS.iter().cloned().collect();
+
+                    let config = ClientConfig::builder()
+                        .with_root_certificates(root_store)
+                        .with_no_client_auth();
+
+                    TlsConnector::from(Arc::new(config))
+                } else {
+                    TlsConnector::from(Arc::new(rustls_platform_verifier::tls_config()))
+                };
+
+                let server_name =
+                    ServerName::try_from(hostname.clone()).map_err(|_| Error::InvalidServerName)?;
+
+                connect_and_start(
+                    config.connect(server_name, stream).await?,
+                    addr,
+                    derp_config.secret_key.clone(),
+                    derp_config.server_keepalives.clone(),
+                    &hostport,
+                    DerpVersion::V2,
+                )
+                .await
+            }
+        };
```

Comment:
This section seems to be mostly copy-pasted from below. Please extract it to avoid duplication.